lib/std/Build/Watch.zig

```diff
@@ -237,6 +237,300 @@ const Os = switch (builtin.os.tag) {
             }
         }
     },
+    .windows => struct {
+        const posix = std.posix;
+        const windows = std.os.windows;
+
+        /// Keyed differently but indexes correspond 1:1 with `dir_table`.
+        handle_table: HandleTable,
+        dir_list: std.ArrayListUnmanaged(*Directory),
+        io_cp: ?windows.HANDLE,
+
+        const HandleTable = std.ArrayHashMapUnmanaged(FileId, ReactionSet, FileId.Adapter, false);
+
+        const FileId = struct {
+            volumeSerialNumber: windows.ULONG,
+            indexNumber: windows.LARGE_INTEGER,
+
+            const Adapter = struct {
+                pub fn hash(self: Adapter, a: FileId) u32 {
+                    _ = self;
+                    var hasher = Hash.init(0);
+                    std.hash.autoHash(&hasher, a);
+                    return @truncate(hasher.final());
+                }
+                pub fn eql(self: Adapter, a: FileId, b: FileId, b_index: usize) bool {
+                    _ = self;
+                    _ = b_index;
+                    return a.volumeSerialNumber == b.volumeSerialNumber and a.indexNumber == b.indexNumber;
+                }
+            };
+        };
+
+        const Directory = struct {
+            handle: windows.HANDLE,
+            id: FileId,
+            overlapped: windows.OVERLAPPED,
+            buffer: [64512]u8 align(@alignOf(windows.FILE_NOTIFY_INFORMATION)) = undefined,
+
+            extern "kernel32" fn ReadDirectoryChangesW(
+                hDirectory: windows.HANDLE,
+                lpBuffer: [*]align(@alignOf(windows.FILE_NOTIFY_INFORMATION)) u8,
+                nBufferLength: windows.DWORD,
+                bWatchSubtree: windows.BOOL,
+                dwNotifyFilter: windows.DWORD,
+                lpBytesReturned: ?*windows.DWORD,
+                lpOverlapped: ?*windows.OVERLAPPED,
+                lpCompletionRoutine: windows.LPOVERLAPPED_COMPLETION_ROUTINE,
+            ) callconv(windows.WINAPI) windows.BOOL;
+
+            fn readChanges(self: *@This()) !void {
+                const notify_filter =
```

Comment:
Future enhancement: this could be made into a packed struct like so:

```zig
pub const FILE_NOTIFY_CHANGE = packed struct(DWORD) {
    FILE_NAME: bool = false,
    DIR_NAME: bool = false,
    ATTRIBUTES: bool = false,
    SIZE: bool = false,
    LAST_WRITE: bool = false,
    LAST_ACCESS: bool = false,
    CREATION: bool = false,
    SECURITY: bool = false,
};
```

Reply:
I can make a separate commit in this PR for that.

Reply:
Just realised that I messed up the order of the fields here. Mingw has the full list [here](https://github.com/mirror/mingw-w64/blob/eff726c461e09f35eeaed125a3570fa5f807f02b/mingw-w64-tools/widl/include/winnt.h#L5162-L5174). You can skip `FILE_NOTIFY_CHANGE_NAME` since its only used to select both `FILE_NAME` and `DIR_NAME`. Finally, don't forget the padding - I forgot to include it.